From 46692e5a1a33343f21d55fa04dd768964d177dfa Mon Sep 17 00:00:00 2001
From: Chad Dupuis <cdupuis@redhat.com>
Date: Thu, 21 Aug 2014 13:05:33 -0400
Subject: [ethernet] qlcnic: Dump mailbox registers when mailbox command times out

Message-id: <1408626422-3941-50-git-send-email-cdupuis@redhat.com>
Patchwork-id: 88579
O-Subject: [PATCH RHEL 7.1 049/138] qlcnic: Dump mailbox registers when mailbox command times out.
Bugzilla: 1089352
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Manish chopra <manish.chopra@qlogic.com>

Bug 1089352 (https://bugzilla.redhat.com/show_bug.cgi?id=1089352)

net commit id 0951c5c214e525185572204b587e3b6762d121f4

Signed-off-by: Manish Chopra <manish.chopra@qlogic.com>
Signed-off-by: Himanshu Madhani <himanshu.madhani@qlogic.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
index 4385fbc..be72f92 100644
--- a/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
+++ b/drivers/net/ethernet/qlogic/qlcnic/qlcnic_83xx_hw.c
@@ -3734,6 +3734,19 @@ static void qlcnic_83xx_decode_mbx_rsp(struct qlcnic_adapter *adapter,
  return;
 }
 
+static inline void qlcnic_dump_mailbox_registers(struct qlcnic_adapter *adapter)
+{
+ struct qlcnic_hardware_context *ahw = adapter->ahw;
+ u32 offset;
+
+ offset = QLCRDX(ahw, QLCNIC_DEF_INT_MASK);
+ dev_info(&adapter->pdev->dev, "Mbx interrupt mask=0x%x, Mbx interrupt enable=0x%x, Host mbx control=0x%x, Fw mbx control=0x%x",
+   readl(ahw->pci_base0 + offset),
+   QLCRDX(ahw, QLCNIC_MBX_INTR_ENBL),
+   QLCRDX(ahw, QLCNIC_HOST_MBX_CTRL),
+   QLCRDX(ahw, QLCNIC_FW_MBX_CTRL));
+}
+
 static void qlcnic_83xx_mailbox_worker(struct work_struct *work)
 {
  struct qlcnic_mailbox *mbx = container_of(work, struct qlcnic_mailbox,
@@ -3778,6 +3791,8 @@ static void qlcnic_83xx_mailbox_worker(struct work_struct *work)
     __func__, cmd->cmd_op, cmd->type, ahw->pci_func,
     ahw->op_mode);
    clear_bit(QLC_83XX_MBX_READY, &mbx->status);
+   qlcnic_dump_mailbox_registers(adapter);
+   qlcnic_83xx_get_mbx_data(adapter, cmd);
    qlcnic_dump_mbx(adapter, cmd);
    qlcnic_83xx_idc_request_reset(adapter,
             QLCNIC_FORCE_FW_DUMP_KEY);
-- 
1.7.1